tortoise/tortoise.go

```diff
@@ -1156,3 +1147,121 @@ func computeExpectedVoteWeight(atxdb atxDataProvider, epochWeight map[types.Epoc
+		for _, atxid := range ballot.EpochData.ActiveSet {
+			atx, err := atxdb.GetAtxHeader(atxid)
+			if err != nil {
+				return nil, fmt.Errorf("atx %s in active set of %s is unknown", atxid, ballot.ID())
+			}
+			total += atx.GetWeight()
+			if atxid == ballot.AtxID {
+				weight = atx.GetWeight()
+				foundWeight = true
+			}
+		}
+		// TODO is it valid not to include your own atxid into active set?
+		if !foundWeight {
+			atx, err := atxdb.GetAtxHeader(ballot.AtxID)
+			if err != nil {
+				return nil, fmt.Errorf("atx %s in active set of %s is unknown", ballot.AtxID, ballot.ID())
+			}
+			weight = atx.GetWeight()
+		}
+
+		expected, err := proposals.GetNumEligibleSlots(weight, total, layerSize, layersPerEpoch)
+		if err != nil {
+			return nil, fmt.Errorf("unable to compute number of eligibile ballots for atx %s", ballot.AtxID)
+		}
+		rst := new(big.Float).SetUint64(weight)
+		rst.Quo(rst, new(big.Float).SetUint64(uint64(expected)))
+		return rst, nil
+	}
+	if ballot.RefBallot == types.EmptyBallotID {
+		return nil, fmt.Errorf("empty ref ballot and no epoch data on ballot %s", ballot.ID())
+	}
+	weight, exist := weights[ballot.RefBallot]
+	if !exist {
+		return nil, fmt.Errorf("ref ballot %s for %s is unknown", ballot.ID(), ballot.RefBallot)
+	}
+	return weight, nil
+}
+
+// recoverBallotWeight recovers all weights from database, instead of persisting weights in tortoise state.
+func recoverBallotWeight(
+	mdb blockDataProvider,
+	atxdb atxDataProvider,
+	opinions map[types.LayerID]map[types.BallotID]Opinion,
+	weights map[types.BallotID]*big.Float,
+	layerSize,
+	layersPerEpoch uint32,
+) error {
+	for _, ballots := range opinions {
+		for ballotID := range ballots {
+			if err := recoverWeightForBallotID(
+				mdb, atxdb, opinions, weights,
+				layerSize, layersPerEpoch, ballotID,
+			); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func recoverWeightForBallotID(
+	mdb blockDataProvider,
+	atxdb atxDataProvider,
+	opinions map[types.LayerID]map[types.BallotID]Opinion,
+	weights map[types.BallotID]*big.Float,
+	layerSize,
+	layersPerEpoch uint32,
+	ballotID types.BallotID,
+) error {
+	block, err := mdb.GetBlock(types.BlockID(ballotID))
+	if err != nil {
+		return fmt.Errorf("failed to load %s from db: %w", ballotID, err)
+	}
+	if block.ATXID == *types.EmptyATXID {
+		return nil
+	}
+
+	ballot := block.ToBallot()
+
+	var refExist bool
+	if ballot.EpochData == nil {
+		_, refExist = weights[ballot.RefBallot]
+		if !refExist {
+			if err := recoverWeightForBallotID(
+				mdb, atxdb, opinions, weights,
+				layerSize, layersPerEpoch, ballot.RefBallot,
+			); err != nil {
+				return err
+			}
+		}
+	}
+	weight, err := computeBallotWeight(atxdb, weights, ballot, layerSize, layersPerEpoch)
+	if err != nil {
+		return err
+	}
+	// reference ballot can be evicted, in order to guarantee that we won't
+	// add evicted ref ballot - we always delete reference weight if it didn't
+	// exist when we computed weight for current ballot
```

Comment:
the logic here seems more complicated than needed. do you think it's worth evicting weight by epoch?

Reply:
i think the current version is simpler overall, indexing weight by epoch will simplify some things and make complex other. this is only for recovery, it is not used in the regular eviction code path